### PR TITLE
Add taskgraph endpoints

### PR DESCRIPTION
**Pull request by snagles**
Rough draft

### CODE REVIEW

**Review by antalakas — APPROVED** [review]

File: openapi-v1.yaml [hunk 1]
```diff
@@ -8400,6 +8462,272 @@ paths:
...
+          required: true
+        - name: graph
+          in: body
+          description: Create the task graph.
+          schema:
+            $ref: "#/definitions/TaskGraph"
+          required: true
+      responses:
+        201:
+          description: Task graph created successfully.
```
Comment (NikolaosPapailiou) [inline openapi-v1.yaml:8640]: Should we also return the id of the TaskGraph?

File: openapi-v1.yaml [hunk 2]
```diff
@@ -8400,6 +8462,272 @@ paths:
...
+            $ref: "#/definitions/TaskGraph"
+        502:
+          description: Bad Gateway
+        default:
+          description: error response
+          schema:
+            $ref: "#/definitions/Error"
+    patch:
+      description: >
+        Update information about a single task graph execution.
```
Comment (NikolaosPapailiou) [inline openapi-v1.yaml:8679]: nit: single task graph

**Review by NikolaosPapailiou — APPROVED** [review]

File: openapi-v1.yaml [hunk 3]
```diff
@@ -8400,6 +8466,210 @@ paths:
...
+            $ref: "#/definitions/TaskGraphLog"
+        502:
+          description: Bad Gateway
+        default:
+          description: error response
+          schema:
+            $ref: '#/definitions/Error'
+
+  # Task graphs
+  /v1/taskgraphs/{namespace}/graphs:
```
Comment (NikolaosPapailiou) [inline openapi-v1.yaml:8530]: v1 is not needed here

**Comment by SarantopoulosKon** [comment]
A few questions @NikolaosPapailiou @snagles 

1. In `TaskGraphLog` definition i don't see any type to get what type the task graph is.
2. Same for `TaskGraph` definition. (no type, if it's batch task graph or not).
3. What's the need for `/taskgraphs/{namespace}/graphs` ? For the UI we use the array browser API for listing `/arrays/browser/owned?page=1&per_page=10&namespace=TileDB-Inc&file_type=registered_task_graph` is it just a convenience endpoint for those that want to see their task graphs programmatically?
4. For the task graph resubmit & stop (e.g. `/taskgraphs/{namespace}/executions/{id}/stop`) the `id` is the task graph log id?
5. We could have "Code preview" only in `TaskGraphNode` (node information for a`RegisteredTaskGraph`) but `TaskGraphNodeMetadata` (node information in task graph logs), have no such information, can we enrich it?
6. In the designs (see image below) we have a `resources` section, can we have that kind of information for the every taskgraph node?

image

**Comment by SarantopoulosKon** [comment]
> > A few questions @NikolaosPapailiou @snagles
> > 
> > 1. In `TaskGraphLog` definition i don't see any type to get what type the task graph is.
> > 2. Same for `TaskGraph` definition. (no type, if it's batch task graph or not).
> > 3. What's the need for `/taskgraphs/{namespace}/graphs` ? For the UI we use the array browser API for listing `/arrays/browser/owned?page=1&per_page=10&namespace=TileDB-Inc&file_type=registered_task_graph` is it just a convenience endpoint for those that want to see their task graphs programmatically?
> > 4. For the task graph resubmit & stop (e.g. `/taskgraphs/{namespace}/executions/{id}/stop`) the `id` is the task graph log id?
> > 5. We could have "Code preview" only in `TaskGraphNode` (node information for a`RegisteredTaskGraph`) but `TaskGraphNodeMetadata` (node information in task graph logs), have no such information, can we enrich it?
> > 6. In the designs (see image below) we have a `resources` section, can we have that kind of information for the every taskgraph node?
> > 
> > image
> 
> Thanks for the feedback @SarantopoulosKon, some comments bellow (@snagles please provide your feedback):
> 
> 1. I think we need to add the TaskGraph type in both  `TaskGraphLog` and `TaskGraph`
> 2. Same as above
> 3. It is a convenience endpoint for now as we haven't consolidated BatchTaskGraphs(/graphs) and RegisteredTaskGraphs. However, I think we need to consolidate these APIs soon.
> 4. Yes this is the log id
> 5. I think the solution here would be to add a `TaskGraphID` to  `TaskGraphLog` and then get the full `TaskGraph` description from `/taskgraphs/{namespace}/graphs/{id}`
> 6. We have `TaskGraphNode.udf_node.environment.resources`

Thanks for the quick reply @NikolaosPapailiou !

Regarding (5), that could solve (6) as well, because `TaskGraphNode.udf_node.environment.resources` is in the registered task graph's node information. The problem is that nodes is an Array which i have to traverse every time to find the correct node in order to get that extra info. If task graphs is really big (thousands of nodes) that will have a performance hit.

If we can't have that info in the log's nodes, what if we had an extra endpoint to get that info? Something like `/taskgraphs/{taskgraph_id}/{client_node_uuid}`

**Review by Shelnutt2 — APPROVED** [review]

**Pull request merged**